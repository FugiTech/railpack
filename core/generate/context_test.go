package generate

import (
	"encoding/json"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/railwayapp/railpack/core/app"
	"github.com/railwayapp/railpack/core/config"
	"github.com/railwayapp/railpack/core/plan"
	"github.com/stretchr/testify/require"
)

type TestProvider struct{}

func (p *TestProvider) Plan(ctx *GenerateContext) error {
	// mise
	mise := ctx.GetMiseStepBuilder()
	nodeRef := mise.Default("node", "18")
	mise.Version(nodeRef, "18", "test")

	// apt
	aptStep := ctx.NewAptStepBuilder("test")
	aptStep.AddAptPackage("git")
	aptStep.AddAptPackage("neofetch")

	// commands
	installStep := ctx.NewCommandStep("install")
	installStep.AddCommand(plan.NewExecCommand("npm install", plan.ExecOptions{}))
	installStep.Outputs = &[]string{"node_modules"}
	installStep.DependsOn = []string{aptStep.Name()}

	buildStep := ctx.NewCommandStep("build")
	buildStep.AddCommand(plan.NewExecCommand("npm run build", plan.ExecOptions{}))
	buildStep.DependsOn = []string{installStep.Name()}

	ctx.Start.Command = "npm run start"

	return nil
}

func CreateTestContext(t *testing.T, path string) *GenerateContext {
	t.Helper()

	userApp, err := app.NewApp(path)
	require.NoError(t, err)

	env := app.NewEnvironment(nil)

	ctx, err := NewGenerateContext(userApp, env)
	require.NoError(t, err)

	return ctx
}

func TestGenerateContext(t *testing.T) {
	ctx := CreateTestContext(t, "../../examples/node-npm")
	provider := &TestProvider{}
	require.NoError(t, provider.Plan(ctx))

	// User defined config
	configJSON := `{
		"packages": {
			"node": "20.18.2",
			"go": "1.23.5",
			"python": "3.13.1"
		},
		"aptPackages": ["curl"],
		"steps": {
			"build": {
				"commands": ["echo building"]
			}
		},
		"secrets": ["RAILWAY_SECRET_1", "RAILWAY_SECRET_2"],
		"start": {
			"variables": {
				"HELLO": "world"
			}
		}
	}`

	var cfg config.Config
	require.NoError(t, json.Unmarshal([]byte(configJSON), &cfg))

	// Apply the config to the context
	require.NoError(t, ctx.ApplyConfig(&cfg))

	// Resolve packages
	resolvedPkgs, err := ctx.ResolvePackages()
	require.NoError(t, err)

	// Generate a plan
	buildOpts := &BuildStepOptions{
		ResolvedPackages: resolvedPkgs,
		Caches:           ctx.Caches,
	}

	// Build and verify each step
	for _, builder := range ctx.Steps {
		_, err := builder.Build(buildOpts)
		require.NoError(t, err)
	}

	buildPlan, _, err := ctx.Generate()
	require.NoError(t, err)

	buildPlanJSON, err := json.Marshal(buildPlan)
	require.NoError(t, err)

	var actualPlan map[string]interface{}
	require.NoError(t, json.Unmarshal(buildPlanJSON, &actualPlan))

	serializedPlan, err := json.MarshalIndent(actualPlan, "", "  ")
	require.NoError(t, err)

	snaps.MatchJSON(t, serializedPlan)
}
