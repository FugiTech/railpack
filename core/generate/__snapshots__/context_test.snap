
[TestGenerateContext - 1]
{
 "caches": {
  "apt": {
   "directory": "/var/cache/apt",
   "type": "locked"
  },
  "apt-lists": {
   "directory": "/var/lib/apt/lists",
   "type": "locked"
  }
 },
 "deploy": {
  "inputs": [
   {
    "step": "build"
   }
  ],
  "startCommand": "echo hello",
  "variables": {
   "HELLO": "world"
  }
 },
 "secrets": [
  "RAILWAY_SECRET_1",
  "RAILWAY_SECRET_2"
 ],
 "steps": [
  {
   "assets": {
    "mise.toml": "[tools]\n  [tools.go]\n    version = \"1.23.5\"\n  [tools.node]\n    version = \"20.18.2\"\n  [tools.python]\n    version = \"3.13.1\"\n"
   },
   "commands": [
    {
     "path": "/mise/shims"
    },
    {
     "customName": "create mise config",
     "name": "mise.toml",
     "path": "/etc/mise/config.toml"
    },
    {
     "cmd": "sh -c 'mise trust -a \u0026\u0026 mise install'",
     "customName": "install mise packages: go, node, python"
    }
   ],
   "inputs": [
    {
     "image": "ghcr.io/railwayapp/railpack-builder-base:latest"
    }
   ],
   "name": "packages:mise",
   "variables": {
    "MISE_CACHE_DIR": "/mise/cache",
    "MISE_CONFIG_DIR": "/mise",
    "MISE_DATA_DIR": "/mise",
    "MISE_INSTALLS_DIR": "/mise/installs",
    "MISE_SHIMS_DIR": "/mise/shims"
   }
  },
  {
   "caches": [
    "apt",
    "apt-lists"
   ],
   "commands": [
    {
     "cmd": "sh -c 'apt-get update \u0026\u0026 apt-get install -y git neofetch'",
     "customName": "install apt packages: git neofetch"
    }
   ],
   "inputs": [
    {
     "step": "packages:mise"
    }
   ],
   "name": "packages:test"
  },
  {
   "commands": [
    {
     "cmd": "npm install"
    }
   ],
   "inputs": [
    {
     "step": "packages:test"
    }
   ],
   "name": "install"
  },
  {
   "commands": [
    {
     "cmd": "sh -c 'echo building'",
     "customName": "echo building"
    }
   ],
   "inputs": [
    {
     "step": "install"
    }
   ],
   "name": "build",
   "secrets": [
    "*"
   ]
  }
 ]
}
---
